### PR TITLE
Fix check 1d

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes https://github.com/matplotlib/matplotlib/issues/22125 and #22330

This was an issue with handling 

```
x = pd.Series(x, dtype="Float32")  
```

in the logic of our plot argument parsing.

`x.values` returns a `FloatingArray`.  But it doesn't behave like a numpy array, in that `x.values[:, None]` returns a `ValueError: values must be a 1D array`.  I actually think this is a Pandas bug, but I'll leave that to someone with more pandas knowledge to inform them.  However, we can get around it by using `x.to_numpy()` which seems to work fine.  

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -1749,11 +1749,12 @@ def test_bar_hatches(fig_test, fig_ref):
 
 def test_pandas_minimal_plot(pd):
     # smoke test that series and index objcets do not warn
-    x = pd.Series([1, 2], dtype="float64")
-    plt.plot(x, x)
-    plt.plot(x.index, x)
-    plt.plot(x)
-    plt.plot(x.index)
+    for x in [pd.Series([1, 2], dtype="float64"),
+              pd.Series([1, 2], dtype="Float32")]:
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py:1753]: Is this specifically for "Float32" or would the test also do with "Float64"? If it's only the custom type and precision does not matter, I'd got with "Float64"  to communicate that we're primarily testing the custom pandas type.

Side-note: It seems the capital "Float" types are not yet documented. There's only [a v1.2.0 change note](https://pandas.pydata.org/docs/whatsnew/v1.2.0.html#experimental-nullable-data-types-for-float-data) and the GH issues linked therein.

Comment (jklymak) [reply lib/matplotlib/tests/test_axes.py:1753]: As far as I can tell its undocumented.  Maybe we should hold of on support, or maybe this PR can go in, but without the test?    

I actually disagree with Pandas having a new type here - it seems people want this for pedantic reasons, but I think 99.999% of the world doesn't care if NaN means a failed computation or missing data.  However, if they feel strongly, they should get numpy onboard, and then everyone will have this flag rather than making a new data type.

File: lib/matplotlib/cbook/__init__.py [hunk 2]
```diff
@@ -1649,7 +1650,7 @@ def index_of(y):
        The x and y values to plot.
     """
     try:
-        return y.index.values, y.values
```
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:1652]: Do we need both this change and the additional exception handling above?

Comment (jklymak) [reply lib/matplotlib/cbook/__init__.py:1652]: I don't know, but I think we should be in the habit of using `to_numpy()` when possible?

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py:1652]: `to_numpy` came in via pandan 0.24 on Jan 25, 2019 so we can rely on it being there and https://stackoverflow.com/a/54508052/380231 makes an arguement in favor of `no_numpy`.

I think this change may be un-related (or fix this by chance) but I do not think it will avoid needing the other one and does no harm.

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/tests/test_axes.py (hunk 1 above)
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py:1753]: Let's just put in
```suggestion
              pd.Series([1, 2], dtype="Float64")]:
```
For clarity and merge. This not code run by the user. The worst thing that can happen is that the test starts failing if numpy changes something.

**Comment by jklymak** [comment]
@tacaswell, thanks for your review, but dismissing as this approach is completely different, and attempts to remove the pandas-ness of the data right at the beginning.

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -1300,47 +1300,16 @@ def _to_unmasked_float_array(x):
...
-            return x
-        # In pandas 1.1.0, multidimensional indexing leads to an
-        # AssertionError for some Series objects, but should be
-        # IndexError as described in
-        # https://github.com/pandas-dev/pandas/issues/35527
-        except (AssertionError, IndexError, TypeError):
+        ndim = x[:, None].ndim
+        if ndim < 2:
             return np.atleast_1d(x)
+        return x
```
Comment (jklymak) [inline lib/matplotlib/cbook/__init__.py:1312]: codecov appears confused here - this return is handled by test_units.py...

**Review by tacaswell — APPROVED** [review]
I have some concerns (mostly because any time we change complicated code we find that the complexity was there to handle some weird corner case that failed to get a test), but overall am :+1: on this as I think this will also fix other data containers.  I think there is a risk that something subtle will break, but we are already dealing with something subtle breaking and I am optimistic that the cure will not be worse than the malady.

If you think of a DF as a 2D array with columns, then this is consistent with how we handle 2D arrays being passed into `plot`.

**Comment by jklymak** [comment]
This gets used in `_plot_args`, usually via `index_of` (which adds a y if it is missing).  

https://github.com/matplotlib/matplotlib/blob/f0593a6daf562d578ec779d92992c10d5b3bd0cb/lib/matplotlib/axes/_base.py#L486-L495

and it gets called _before_ we check units.  So I guess what I am proposing here would strip an object that has a `to_numpy` method but somehow was being used by somebody for units.  

This does give me pause that we have maybe done the wrong thing here, and are using `_check_1d` to both coerce a singleton to an array _and_ to convert pandas series and dataframes to arrays.  Maybe the _right_ thing to do is just return the DataFrame, run through the converter, and _then_ coerce to numpy if it is not already.

File: lib/matplotlib/cbook/__init__.py [hunk 4]
```diff
@@ -1300,47 +1300,16 @@ def _to_unmasked_float_array(x):
...
-            # object.
-            if ndim < 2:
-                return np.atleast_1d(x)
-            return x
-        # In pandas 1.1.0, multidimensional indexing leads to an
-        # AssertionError for some Series objects, but should be
-        # IndexError as described in
-        # https://github.com/pandas-dev/pandas/issues/35527
-        except (AssertionError, IndexError, TypeError):
+        ndim = x[:, None].ndim
```
Comment (dstansby) [inline lib/matplotlib/cbook/__init__.py:1309]: Could you add a comment about what the slicing is doing here? Looking at this with only the context of the surrounding code I'm a bit lost.

Comment (jklymak) [reply lib/matplotlib/cbook/__init__.py]: Well good catch - this wasn't needed at all, at least empirically.  That simplifies this even more...

**Review by dstansby — APPROVED** [review]

**Comment by jklymak** [comment]
@tacaswell, this simplified a bit more since your approval - if you wanted to double check, that would be appreciated.

**Comment by oscargus** [comment]
In #22560 I added `cbook._unpack_pandas` (so `_unpack_pandas` for this purpose) which basically does `to_numpy()`, but also with a fallback to `values`. I did not touch the lines you changed here though.

I guess it can make sense to have a coordinates merge of this and that PR. If this is merged first, I'll update my PR. If my PR is merged first, it can make sense to use that function here.

**Comment by jklymak** [comment]
Well first it's not only pandas that has to_numpy so that is a bit of a misnomer.  But also, why have a separate method at all?

**Comment by oscargus** [comment]
It was suggested to use a separate function. Right now, slightly different approaches are used at different locations in the code. Sometimes a fallback to `values`, sometimes not. A benefit is possibly that it is enough to modify it in a single location (and that old Pandas versions are still supported with this fallback). Also, if we want to support new libraries that with some other name, it will be easy to do that. Well, all the standard reasons to factor out a piece of common code...

Regarding naming, I considered that, but I do not know which other libraries support that function. However, I do not really see the name neither written in stone nor something that should prohibit using a single function, there will be a name that is correct enough. (And as you can see, there are explicit comments mentioning pandas at all the other locations where it was used.)

**Comment by oscargus** [comment]
I then assume that we merge this first, find a good name for the function and, if you want to strongly object using the function here, you can do that in #22560.

**Comment by jklymak** [comment]
I meant why have a separate method than check_1d?   Our problem is inconsistent duckttping so if we can have it all in one spot that would be very helpful.  If check1d does more than duck type pandas then sure it could call the ducktype converter.  

I do actually wonder if all of this should just be part of the unit conversion machinery rather than cbook calls

**Comment by oscargus** [comment]
> I meant why have a separate method than check_1d?

I do not have an enough overview of the code base to see if one should/could have used check_1d (or check_2d?) instead. But if possible, that is of course even better.
